Fix addition of undesirable white space in HTML

jwierzbi opened the pull request:
Hi,

First of all let me say that this is an amazing plugin. Thank you for putting an effort to create it :blush:.

Second I'm not very experienced with the GitHub workflow so apologies if I messed something up here :sweat_smile:.

I started using your plugin recently and I noticed an issue with some additional white space added in my HTML output, e.g. a simple paragraph like so:

    To create the pending node we use the :code:`nodes.pending()` method and pass 3 arguments to it...

would turn into following HTML output:
  
      <p>
            To create the pending node we use the
            <code>
             nodes.pending()
            </code>
            method and pass
      3 arguments to it...

which doesn't look very nice. This instead should be:

    <p>To create the pending node we use the <code>nodes.pending()</code> method and pass 3 arguments to it...

I narrowed down the problem to the `html_file.write(soup.prettify())` line in *pelican/plugins/seo/seo_enhancer/\__init__.py*.

I think I figured out a potential fix for this issue hence this pull request. Please let me know if you consider this is/isn't a real problem and if this fix would be of any use to you.

Best regards,
jwierzbi

Commit message:

    BeautifulSoup.prettify() changes formatting for the HTML output which
    leads to addition of undesirable and ugly white spaces to inline
    elements like code (<code>) or anchor (<a>) elements.
    
    The solution is to user 'str()' on the soup object so that the HTML
    content is not changed.

MaevaBrunelles commented:
Hi @jwierzbi,

I'm really glad to read that you're enjoying the plugin 😊

You're raising an interesting subject. The real issue here is that my plugin steps in after Pelican generates the HTML file by reading it, playing with it, and re-writing it. So we loose the Jinja template HTML indentation.
That's why initially, I chose to prettify the written HTML to keep a semblance of indentation, but I agree it's not perfect.

In my opinion, the solution you proposed is not perfect either because the HTML generated will not have indentation at all, which is not convenient:
```html
<!DOCTYPE html>

<html lang="fr">
<head>
<meta charset="utf-8"/>
<meta content="Pelican" name="generator"/>
<title>My First Review</title>
```

Food for thought:
- have a plugin option to let the user choose between prettify or not
- stop using `bs4` which breaks the HTML indentation and find another way to add HTML to the generated file

@justinmayer, @avaris: would you mind telling us your opinion on this subject? Is it a good or a bad practice for a plugin to rewrite the HTML like I currently do? I'm not sure if there is another way of doing as I need to intervene after the HTML generation. Or maybe this is not a subject and we don't mind to break the template indentation?

jwierzbi commented:
Hi @MaevaBrunelles,

> In my opinion, the solution you proposed is not perfect either because the HTML generated will not have indentation at all ...

Fair enough. Silly question but I really don't know much about SEO. Does the indentation has any influence on SEO or is it purely for readability reasons? I'm asking because in addition to your plugin I'm also using css-html-js-minify.

This I guess brings another topic which probably isn't a big deal but might be significant in some cases. Beautifying beside affecting the visual side of the output also increases HTML file sizes. Not sure if this is a parameter that influences SEO in any way though.

> have a plugin option to let the user choose between prettify or not

I probably should have done this in the first place :sweat_smile:. I went ahead and added a setting for enabling/disabling pretty printing and updated the pull request. Can you please take a look if looks better then before. Fingers crossed :wink:.

> stop using bs4 which breaks the HTML indentation and find another way to add HTML to the generated file

This seems like a best long term solution but it's a little bit above my knowledge about HTML parsing, etc the moment :sweat_smile:. I did a quick test with lxml which apparently supports pretty printing but it's not perfect when it comes to indentation as well. It keeps the inline tags unchanged though (at least it didn't change those in my simple test scenario).

avaris commented:
I wouldn't consider this "good" or "bad" practice. This is pretty much a limitation of parse-unparse cycle where parser strips away useless(!) information like whitespace or comments. Personally, I'd make sure HTTP compression is properly setup in the server and then not worry about whitespaces in HTML that much :).

justinmayer commented:
> […] which doesn't look very nice. This instead should be: […]

I'm not sure I agree that what `soup.prettify()` did here is undesirable. Sure, it could have determined that a short value wrapped in `<code>[…]</code>` should be in-line, but that's not an easy call to make, and I think its decision to make it look like a code block is arguably the most reasonable thing to do with these cases.

Regarding HTML size, I agree with @avaris that whitespace isn't an issue because properly-configured web servers should compress HTML before sending to the browser.

Perhaps [Prettier](https://prettier.io) or [HTML Tidy](https://www.html-tidy.org) would do a better job at auto-formatting HTML, but I'm not sure that theoretical improvement would be worth the extra dependency.

jwierzbi commented:
Hi,

> I'm not sure I agree that what soup.prettify() did here is undesirable.

I probably should have done a better job explaining the issue that I can see. Let's take a simple article like so:

![Screenshot_2021-04-27 Test Site](https://user-images.githubusercontent.com/23691977/116202902-ae270a00-a76d-11eb-9544-79efde9da4c8.png)

When the HTML output goes through `soup.prettify()` something like below is produces:

![Screenshot_2021-04-27 Test Article](https://user-images.githubusercontent.com/23691977/116203005-cd259c00-a76d-11eb-8519-fb0c56a682b1.png)

Notice the extra spaces between brackets and the link and the extra space after the `code`. So this actually affects how the content is displayed. I don't think this should be ignored especially since this affects not only `<code>` but also other inline elements.

I agree with the arguments about the size. It was merely a concern from my side but it seems that it was unnecessary. Thanks for clarifying that for me :relieved:.

Hope this makes sense :sweat_smile:.

justinmayer commented:
That does make more sense. Perhaps update the change-log entry in `RELEASE.md` to somehow reflect (1) neither is good/bad and thus not really a "fix" and (2) the fact that this is now a setting?

MaevaBrunelles commented:
Indeed, if `soup.prettify()` has an impact on how the HTML is rendered, it's an issue. So I vote for the initial solution proposed by @jwierzbi, which is simply to do a `str(soup)`. I'll drop the idea of a setting as we don't really mind of how the HTML is generated (and as it's hard to to better).

MaevaBrunelles commented on pelican/plugins/seo/tests/conftest.py:
```diff
@@ -193,6 +193,52 @@ def fake_article_multiple_elements():
     )
 
 
+@pytest.fixture()
```
You don't have to create a new fixture for your need. Just update https://github.com/pelican-plugins/seo/blob/master/pelican/plugins/seo/tests/conftest.py#L61 by adding the HTML code you need. Of course you'll need to update the test which use this fixture too.

jwierzbi replied:
Removed the new test fixture and added additional content to the `fake_article` fixture.

MaevaBrunelles commented on pelican/plugins/seo/tests/test_html_content.py:
```diff
@@ -0,0 +1,40 @@
+""" Units tests for SEO Enhancer. """
```
I guess that with my previous comment, you don't need to add a test. All the tests in [this file](https://github.com/pelican-plugins/seo/blob/master/pelican/plugins/seo/tests/test_seo_enhancer.py) you'll have to update will test if no whitespace is inserted.

jwierzbi replied:
Removed the new test file and updated test cases in _test_seo_enhancer.py_. I used BeautifulSoup to format the output so the tests look more readable but added additional asserts on raw output.

jwierzbi replied:
I was going for the minimal change but I agree that reformatting the HTML is the right thing to do. Fixed.

MaevaBrunelles commented on RELEASE.md:
```diff
@@ -0,0 +1,3 @@
+Release type: patch
+
+Fix addition of undesirable white space in HTML
```
You can update the release message as @justinmayer proposed in its (1) point.

jwierzbi replied:
Done. Not sure it this is enough. I looked at the previous files and they usually have one line summaries so I tried to do the same. I can I more detailed description if required.

MaevaBrunelles replied:
It looks better! I would just add the root cause, for example: `Fix addition of undesirable white space in HTML because of BeautifulSoup.prettiy()`

jwierzbi replied:
Thank you for the suggestion. Changed.

MaevaBrunelles commented on pelican/plugins/seo/tests/test_seo_enhancer.py:
```diff
@@ -94,7 +95,7 @@ def test_add_html_enhancements_to_file(self, fake_article, fake_seo_enhancer):
 
             assert (
-                fake_html_content
+                BeautifulSoup(fake_html_content, features="html.parser").prettify()
```
```suggestion
               fake_html_content
```

jwierzbi replied:
Fixed.

MaevaBrunelles commented on pelican/plugins/seo/tests/test_html_content.py:
```diff
@@ -0,0 +1,40 @@
+            fake_html_content = write_args[0]
+
+            assert "<code>an inline code</code>" in fake_html_content
+            assert '"<a href="http://cool.site/">cool site</a>"' in fake_html_content
```
```suggestion
```

jwierzbi replied:
This file should not be in the change anymore, but I removed the additional asserts that I added in test_seo_enhancer.py.

MaevaBrunelles commented on pelican/plugins/seo/tests/test_seo_enhancer.py:
```diff
@@ -2,6 +2,7 @@
 
 from unittest.mock import mock_open, patch
 
+from bs4 import BeautifulSoup
```
```suggestion
```

jwierzbi replied:
Removed.

jwierzbi commented on pelican/plugins/seo/tests/test_seo_enhancer.py:
```diff
@@ -152,7 +153,7 @@ def test_add_html_enhancements_to_file_with_open_graph(
 
             assert (
-                fake_html_content
+                BeautifulSoup(fake_html_content, features="html.parser").prettify()
```
Fixed.

jwierzbi commented on pelican/plugins/seo/tests/test_seo_enhancer.py:
```diff
@@ -222,7 +223,7 @@ def test_add_html_enhancements_to_file_with_twitter_cards(
 
             assert (
-                fake_html_content
+                BeautifulSoup(fake_html_content, features="html.parser").prettify()
```
Fixed.

MaevaBrunelles commented:
@jwierzbi your patch is available in the 1.2.1 version, many thanks for your contribution!
@avaris @justinmayer thanks for taking time to give your inputs :)

jwierzbi commented:
@MaevaBrunelles That's great. Thank you very much for taking the time to review it and guiding me through the whole process :smile:.
